source/extensions/common/dynamic_forward_proxy/dns_cache.h

```diff
@@ -43,6 +44,21 @@ class DnsHostInfo {
 
 using DnsHostInfoSharedPtr = std::shared_ptr<DnsHostInfo>;
 
+/**
+ * A resource manager of DNS Cache.
+ */
+class DnsCacheResourceManager {
+public:
+  virtual ~DnsCacheResourceManager() = default;
+
+  /**
+   * Returns the resource limit of pending requests to DNS.
+   */
+  virtual ResourceLimit& pendingRequests() PURE;
+};
+
+using DnsCacheResourceManagerPtr = std::shared_ptr<DnsCacheResourceManager>;
```

Comment:
style nit: `DnsCacheResourceManagerSharedPtr`

second thought: does this have to be a shared_ptr? from the usage below it seems a unique_ptr is enough.